# Exercises: Heuristic Search Exercises

## Instructions
- Answer all questions in English.
- Show your reasoning clearly for theoretical questions.
- For programming tasks, write clean and well-documented Python code.
- You may use only the Python standard library unless otherwise stated.

## Learning Focus
This exercise set focuses on:
1. Heuristic search algorithms
2. Methods for constructing heuristic functions
3. Solving a problem programmatically using heuristic search

## Suggested Time Allocation
- Part A: Conceptual questions (25 minutes)
- Part B: Heuristic design questions (25 minutes)
- Part C: Programming task (40 minutes)


## Part A - Heuristic Search Algorithms

### Exercise 1. Greedy Best-First Search vs A* Search

Answer the following:

1. What is the evaluation function used in Greedy Best-First Search?
2. What is the evaluation function used in A* Search?
3. What is the role of `g(n)` in A* Search?
4. Why can Greedy Best-First Search fail to find an optimal solution?
5. Why can A* Search still reconsider paths that Greedy Best-First Search may ignore?


#### 1. What is the evaluation function used in Greedy Best-First Search?

The evaluation function used in Greedy Best-First Search is typically a heuristic function, denoted as h(n), which estimates the cost to reach the goal from node n. The algorithm selects the node with the lowest h(n) value to expand next, aiming to find a path to the goal as quickly as possible.

#### 2. What is the evaluation function used in A* Search?

The evaluation function used in A* Search is typically denoted as f(n) and is defined as the sum of two components: g(n) and h(n).
- g(n) represents the cost from the start node to the current node n. It is the actual cost incurred to reach node n from the starting point.
- h(n) represents the heuristic estimate of the cost from node n to the goal node. It is an estimate of the remaining cost to reach the goal from node n.

Therefore, the evaluation function can be expressed as:
**f(n) = g(n) + h(n)**

In A* Search, the algorithm selects the node with the lowest f(n) value for expansion, which allows it to efficiently find the optimal path to the goal while considering both the cost incurred so far and the estimated cost to reach the goal.

#### 3. What is the role of `g(n)` in A* Search?

`g(n)` represents the cost from the start node to the current node `n`. It is used in A* Search to calculate the total estimated cost `f(n)`, which is the sum of `g(n)` and `h(n)`, where `h(n)` is the heuristic estimate of the cost from node `n` to the goal. 

The A* algorithm uses `f(n)` to prioritize which nodes to explore next, aiming to find the most efficient path to the goal.

#### 4. Why can Greedy Best-First Search fail to find an optimal solution?

Greedy Best-First Search can fail to find an optimal solution because it relies solely on a heuristic function to guide its search. The algorithm selects the node that appears to be closest to the goal based on the heuristic, without considering the actual cost of reaching that node from the start. 

This can lead to situations where the algorithm gets stuck in a local minimum or chooses a path that seems promising but ultimately leads to a suboptimal solution. Additionally, if the heuristic is not well-designed or is misleading, it can further exacerbate this issue, causing the algorithm to overlook better paths that may have a higher initial cost but ultimately lead to an optimal solution.

#### 5. Why can A* Search still reconsider paths that Greedy Best-First Search may ignore?

A* Search can reconsider paths that Greedy Best-First Search may ignore because A* Search uses a combination of the actual cost from the start node to the current node (g(n)) and the heuristic estimate of the cost from the current node to the goal (h(n)). 

This allows A* Search to evaluate paths based on both their actual cost and their estimated cost to reach the goal, which can lead to reconsidering paths that may have been ignored by Greedy Best-First Search, which only considers the heuristic estimate (h(n)) and may overlook paths that have a higher heuristic value but a lower actual cost.

### Exercise 2. Completeness and Optimality

Fill in the table below and briefly justify each answer.

| Algorithm | Complete? | Optimal? | Short justification |
|---|---|---|---|
| Greedy Best-First Search |  |  |  |
| A* Search |  |  |  |
| Weighted A* Search (W > 1) |  |  |  |

Your explanation should mention the effect of the heuristic and whether optimality is guaranteed.

| Algorithm | Complete? | Optimal? | Short justification |
|---|---|---|---|
| Greedy Best-First Search | No | No | Greedy Best-First Search is not complete because it can get stuck in loops or fail to find a solution if the heuristic is misleading. It is not optimal because it does not consider the cost of the path taken, only the estimated cost to the goal. |
| A* Search | Yes | Yes | A* Search is complete because it will eventually explore all possible paths if a solution exists. It is optimal because it uses both the cost to reach the current node and the estimated cost to reach the goal, ensuring that it finds the least costly path. |
| Weighted A* Search (W > 1) | Yes | No | Weighted A* Search is complete because it will explore all paths if a solution exists. However, it is not optimal because the weight (W > 1) can cause the algorithm to prioritize paths that are not the least costly, leading to suboptimal solutions.|

### Exercise 3. Interpreting the Evaluation Function

Suppose a node `n` has:
- `g(n) = 7`
- `h(n) = 5`

Answer the following:

1. What is `f(n)` for A* Search?
2. What value would Greedy Best-First Search use to rank this node?
3. If another node `m` has `g(m) = 3` and `h(m) = 8`, which node would A* prefer?
4. Which node would Greedy Best-First Search prefer?
5. Explain why the two algorithms may make different choices.

#### 1. What is `f(n)` for A* Search?
`f(n)` is calculated as the sum of `g(n)` and `h(n)`.

Therefore, `f(n) = g(n) + h(n) = 7 + 5 = 12`.

#### 2. What value would Greedy Best-First Search use to rank this node?

Greedy Best-First Search uses the heuristic function `h(n)` to rank nodes. Therefore, it would use the value of `h(n)`, which is 5, to rank this node.

#### 3. If another node `m` has `g(m) = 3` and `h(m) = 8`, which node would A* prefer?

A* would prefer node `m` because it has a lower total cost `f(m) = g(m) + h(m) = 3 + 8 = 11` compared to node `n` which has `f(n) = g(n) + h(n) = 7 + 5 = 12`.

#### 4. Which node would Greedy Best-First Search prefer?

Greedy Best-First Search would prefer the node with the lowest heuristic value `h(n)`. In this case, it would prefer node `n` because `h(n) = 5`, which is lower than any other node's heuristic value.

#### 5. Explain why the two algorithms may make different choices.

The two algorithms, A* and Greedy Best-First Search, may make different choices because they use different evaluation functions to determine which node to expand next.

- A* Search uses the evaluation function `f(n) = g(n) + h(n)`, which combines the cost to reach the node (`g(n)`) and the estimated cost to reach the goal from that node (`h(n)`). In this case, `f(n) = 7 + 5 = 12`.
- Greedy Best-First Search, on the other hand, uses the evaluation function `f(n) = h(n)`, which only considers the estimated cost to reach the goal from that node. In this case, `f(n) = 5`.

Because A* considers both the cost to reach the node and the estimated cost to reach the goal, it may choose to expand a different node that has a lower total cost (`f(n)`) compared to Greedy Best-First Search, which only considers the heuristic value (`h(n)`). This can lead to different paths being explored and potentially different solutions being found by the two algorithms.

### Exercise 4. Weighted A* Search

Weighted A* uses the evaluation function:

`f(n) = g(n) + W * h(n)`

Answer the following:

1. What happens when `W = 1`?
2. What happens when `W = 0`?
3. What happens conceptually when `W` becomes very large?
4. Why can Weighted A* often expand fewer nodes?
5. Why does Weighted A* lose the optimality guarantee when `W > 1`?


#### 1. What happens when `W = 1`?

When `W = 1`, the evaluation function simplifies to:
`f(n) = g(n) + h(n)`

This is the standard A* search algorithm, which combines the actual cost from the start node to the current node (`g(n)`) with the heuristic estimate of the cost from the current node to the goal node (`h(n)`). In this case, the search will be optimal and complete, as it will always expand the node with the lowest total cost.

#### 2. What happens when `W = 0`?

When `W = 0`, the evaluation function simplifies to `f(n) = g(n)`. This means that the algorithm will only consider the cost from the start node to the current node (g(n)) and will ignore the heuristic estimate of the cost to reach the goal (h(n)). 

As a result, the algorithm will behave like Dijkstra's algorithm, which explores all possible paths based on their actual costs without any guidance from a heuristic. This can lead to inefficient search, especially in large search spaces, as it may explore many unnecessary nodes before reaching the goal.

#### 3. What happens conceptually when `W` becomes very large?

When `W` becomes very large, the evaluation function `f(n)` becomes dominated by the heuristic component `W * h(n)`. This means that the search algorithm will heavily prioritize nodes that are estimated to be closer to the goal, as indicated by the heuristic function `h(n)`. As a result, the search may become more greedy, potentially leading to faster solutions in some cases. 

However, it can also lead to suboptimal solutions or even failure to find a solution if the heuristic is not accurate, as the algorithm may overlook paths that have a lower actual cost `g(n)` but a higher heuristic estimate. In extreme cases, if `W` is excessively large, the algorithm may behave like a pure greedy search, which can be inefficient and may not guarantee finding the optimal solution.

#### 4. Why can Weighted A* often expand fewer nodes?

Weighted A* can often expand fewer nodes because it places more emphasis on the heuristic estimate (h(n)) by multiplying it with a weight (W). This means that the algorithm is more likely to prioritize nodes that are estimated to be closer to the goal, potentially leading to a faster path to the solution. By focusing on promising paths, Weighted A* can avoid exploring less relevant nodes, thus reducing the total number of nodes expanded.

#### 5. Why does Weighted A* lose the optimality guarantee when `W > 1`?

Weighted A* loses the optimality guarantee when `W > 1` because it places more emphasis on the heuristic estimate `h(n)` than on the actual cost `g(n)`. When `W` is greater than 1, the algorithm may prioritize nodes that appear to be closer to the goal based on the heuristic, even if they have a higher actual cost. 

This can lead to situations where the algorithm selects a suboptimal path that seems promising due to the inflated heuristic value, rather than exploring paths that may have a lower total cost. As a result, Weighted A* with `W > 1` may not always find the optimal solution.

## Part B - Designing Heuristic Functions

### Exercise 5. Admissible Heuristics

A heuristic `h` is admissible if it never overestimates the true cost to the goal.

For each statement below, write **True** or **False** and explain why.

1. Every admissible heuristic is guaranteed to be exact.
2. If `h(n) = 0` for all nodes, then `h` is admissible.
3. A heuristic that sometimes overestimates the true cost can still guarantee optimal A* solutions.
4. Straight-line distance is often admissible for route-finding problems on road maps.
5. If `h(n)` is admissible, A* Search is optimal.


#### 1. Every admissible heuristic is guaranteed to be exact.

`False`. An admissible heuristic is one that never overestimates the true cost to reach the goal, but it can still underestimate the cost. An exact heuristic would always provide the true cost, while an admissible heuristic can be less than or equal to the true cost, but not greater.

#### 2. If `h(n) = 0` for all nodes, then `h` is admissible.

`True`. If `h(n) = 0` for all nodes, it means that the heuristic estimates the cost to reach the goal as zero for every node. Since the true cost to reach the goal can never be negative, this heuristic will never overestimate the true cost. Therefore, it is admissible.

#### 3. A heuristic that sometimes overestimates the true cost can still guarantee optimal A* solutions.

`False`. A heuristic that overestimates the true cost can lead A* to make suboptimal choices, as it may prioritize paths that appear cheaper than they actually are, potentially missing the optimal solution.

#### 4. Straight-line distance is often admissible for route-finding problems on road maps.

`True`. Straight-line distance (also known as the Euclidean distance) is often admissible for route-finding problems on road maps because it represents the shortest possible distance between two points. 

Since actual travel paths on roads may involve turns, detours, and other factors that can increase the distance, the straight-line distance will never overestimate the true cost to reach the goal. Therefore, it is an admissible heuristic in this context.

#### 5. If `h(n)` is admissible, A* Search is optimal.

`True`. A* Search is optimal when using an admissible heuristic because it guarantees that the estimated cost to reach the goal (h(n)) will never exceed the actual cost. This means that A* will always find the least costly path to the goal, ensuring optimality.

### Exercise 6. Heuristics from Relaxed Problems

Consider the 8-puzzle.

Two well-known heuristics are:
- `h1(n)`: number of misplaced tiles
- `h2(n)`: total Manhattan distance

Answer the following:

1. What is a relaxed problem?
2. What relaxation leads to the misplaced-tiles heuristic?
3. What relaxation leads to the Manhattan-distance heuristic?
4. Why does solving a relaxed problem provide an admissible heuristic?
5. Between `h1` and `h2`, which one is usually more informative? Why?

#### 1. What is a relaxed problem?

A relaxed problem is a simplified version of the original problem where certain constraints are removed. In the context of the 8-puzzle, a relaxed problem could be one where we ignore the rule that only one tile can be moved at a time, allowing multiple tiles to be moved simultaneously. 

This relaxation can help in estimating the cost to reach the goal state, as it provides a lower bound on the number of moves required.

#### 2. What relaxation leads to the misplaced-tiles heuristic?

The misplaced-tiles heuristic (h1) can be derived from a relaxation of the 8-puzzle problem where we ignore the specific rules of tile movement and simply count the number of tiles that are out of place compared to the goal state. 

In this relaxed version, we assume that any tile can be moved directly to its correct position without considering the constraints of the puzzle, such as the empty space or the need to move other tiles out of the way. 

This allows us to calculate the heuristic by simply counting how many tiles are not in their correct positions, which gives us an estimate of how many moves are needed to solve the puzzle.

#### 3. What relaxation leads to the Manhattan-distance heuristic?

The relaxation that leads to the Manhattan-distance heuristic is to allow tiles to move through each other without any restrictions. In this relaxed version of the problem, the cost of moving a tile is equal to the Manhattan distance from its current position to its goal position, regardless of any other tiles in the way. 

This allows us to calculate the heuristic value based on the total Manhattan distance of all tiles from their goal positions, which provides an estimate of the minimum number of moves required to solve the puzzle.

#### 4. Why does solving a relaxed problem provide an admissible heuristic?

Solving a relaxed problem provides an admissible heuristic because it simplifies the original problem by removing certain constraints, making it easier to solve. The solution to the relaxed problem will always be less than or equal to the cost of the optimal solution to the original problem, ensuring that the heuristic does not overestimate the true cost. 

This is what makes it admissible, as it guarantees that the heuristic will never overestimate the cost to reach the goal state.

#### 5. Between `h1` and `h2`, which one is usually more informative? Why?

`h2(n)`, the total Manhattan distance, is usually more informative than `h1(n)`, the number of misplaced tiles. This is because `h2(n)` takes into account not only whether a tile is misplaced but also how far it is from its goal position. 

In contrast, `h1(n)` simply counts the number of misplaced tiles without considering their distance from the correct position. Therefore, `h2(n)` provides a more accurate estimate of the cost to reach the goal state, making it a more effective heuristic for guiding search algorithms in solving the 8-puzzle.

### Exercise 7. Heuristics from Subproblems

Answer the following questions about subproblem-based heuristics:

1. What does it mean to build a heuristic from a subproblem?
2. Why can the exact cost of solving a subproblem be used as an admissible heuristic for the full problem?
3. What is a pattern database?
4. Why are pattern databases useful in search problems such as the 8-puzzle?
5. Give one advantage and one limitation of subproblem-based heuristics.


#### 1. What does it mean to build a heuristic from a subproblem?

Building a heuristic from a subproblem involves creating an approximation or a simplified version of the original problem to guide the search for a solution. 

This is often done by identifying a smaller, more manageable portion of the problem that can be solved more easily, and then using the insights gained from that subproblem to inform the decision-making process in the larger problem. 

The heuristic can help prioritize certain paths or actions based on the results obtained from solving the subproblem, ultimately leading to a more efficient search for the optimal solution in the original problem.

#### 2. Why can the exact cost of solving a subproblem be used as an admissible heuristic for the full problem?

The exact cost of solving a subproblem can be used as an admissible heuristic for the full problem because it provides a lower bound on the cost to reach the goal from the current state. 

An admissible heuristic is one that never overestimates the true cost to reach the goal, and since the exact cost of solving a subproblem is a precise measure of the cost to solve that specific part of the problem, it will always be less than or equal to the actual cost of solving the entire problem. 

Therefore, using the exact cost of a subproblem as a heuristic ensures that it remains admissible and can effectively guide search algorithms towards optimal solutions.

#### 3. What is a pattern database?

A pattern database is a precomputed table that stores the optimal solutions or heuristic values for specific subproblems of a larger problem. 

It is often used in search algorithms to improve efficiency by providing quick access to previously computed results, allowing the algorithm to make informed decisions and avoid redundant calculations. 

Pattern databases are commonly used in combinatorial optimization problems, such as puzzles and games, where they can significantly reduce the search space and improve performance.

#### 4. Why are pattern databases useful in search problems such as the 8-puzzle?

Pattern databases are useful in search problems like the 8-puzzle because they provide precomputed heuristic values that can significantly speed up the search process. A pattern database is a lookup table that contains the optimal number of moves required to solve a specific subproblem of the larger problem. 

In the case of the 8-puzzle, a pattern database can store the minimum number of moves required to arrange a subset of the tiles in their correct positions.By using a pattern database, the search algorithm can quickly retrieve heuristic values for different states of the puzzle, allowing it to make informed decisions about which paths to explore. 

This can lead to a significant reduction in the number of states that need to be evaluated, ultimately improving the efficiency of the search and reducing the time it takes to find a solution.

#### 5. Give one advantage and one limitation of subproblem-based heuristics.

**Advantage**: Subproblem-based heuristics can break down a complex problem into smaller, more manageable parts, allowing for more efficient problem-solving and potentially faster solutions.

**Limitation**: However, subproblem-based heuristics may not always capture the global optimal solution,as they focus on solving individual subproblems without considering the overall problem context, which can lead to suboptimal results.


### Exercise 8. Dominance of Heuristics

Suppose `h1` and `h2` are both admissible heuristics, and for every node `n`,

`h2(n) >= h1(n)`

Answer the following:

1. What does it mean to say that `h2` dominates `h1`?
2. Why does A* with `h2` usually expand fewer nodes than A* with `h1`?
3. Does dominance mean that `h2` is always exact? Explain.
4. If `h2` is larger than `h1` only for some nodes, can it still be useful? Why?
5. Explain the role of `C*` in the analysis of A* node expansion.


#### 1. What does it mean to say that `h2` dominates `h1`?

It means that `h2` is a better heuristic than `h1` because it provides a higher estimate of the cost to reach the goal from any given node. In other words, `h2` is more informed and can potentially lead to a more efficient search process compared to `h1`.

#### 2. Why does A* with `h2` usually expand fewer nodes than A* with `h1`?

A* with `h2` usually expands fewer nodes than A* with `h1` because `h2` provides a more accurate estimate of the cost to reach the goal. Since `h2(n) >= h1(n)` for every node `n`, it means that `h2` is a more informed heuristic compared to `h1`. 

When A* uses `h2`, it can prioritize nodes that are closer to the goal more effectively, leading to a more efficient search. In contrast, A* with `h1` may expand more nodes because it has a less accurate estimate, causing it to explore paths that are not as promising. 

Therefore, A* with `h2` typically results in fewer node expansions and a faster pathfinding process.

#### 3. Does dominance mean that `h2` is always exact? Explain.

`No`, dominance does not necessarily mean that `h2` is always exact. An admissible heuristic is one that never overestimates the true cost to reach the goal, but it can still underestimate it. 

If `h2` dominates `h1`, it means that `h2` provides a higher estimate than `h1` for every node, but it does not guarantee that `h2` will always provide the exact cost to reach the goal. 

Therefore, while `h2` may be more informed than `h1`, it can still be an underestimate and thus not exact.

#### 4. If `h2` is larger than `h1` only for some nodes, can it still be useful? Why?

`Yes`, `h2` can still be useful even if it is larger than `h1` only for some nodes. This is because the effectiveness of a heuristic in guiding the search process depends on how well it estimates the cost to reach the goal from a given node. 

If `h2` provides a better estimate (i.e., is larger) for certain nodes, it can help the search algorithm prioritize those nodes more effectively, potentially leading to faster convergence towards the goal. In contrast, if `h1` is consistently lower than `h2`, it may underestimate the cost to reach the goal, which could lead to exploring more nodes than necessary. 

Therefore, even if `h2` is not universally better than `h1`, it can still be a valuable heuristic for guiding the search process in specific cases.

#### 5. Explain the role of `C*` in the analysis of A* node expansion.

`C*` represents the cost of the optimal solution path from the start node to the goal node. In the analysis of A* node expansion, `C*` is used as a benchmark to evaluate the efficiency of the heuristic function. 

A* expands nodes based on the evaluation function `f(n) = g(n) + h(n)`, where `g(n)` is the cost from the start node to node `n`, and`h(n)` is the heuristic estimate of the cost from node `n` to the goal. 

If the heuristic is admissible, it will never overestimate the true cost to reach the goal, meaning that `h(n) <= C* - g(n)` for all nodes `n`. This ensures that A* will expand nodes in a way that leads to an optimal solution, as it will not overlook any potentially optimal paths. 

The value of `C*` helps to determine how many nodes A* will expand before finding the optimal solution, as it provides a threshold for the heuristic estimates.


### Exercise 9. Combining Heuristics

Suppose you have three admissible heuristics: `h1`, `h2`, and `h3`.

1. Write the formula for combining them into a single heuristic that remains admissible.
2. Why is this combination admissible?
3. Why is taking the maximum usually better than choosing only one fixed heuristic?
4. Would taking the sum `h1(n) + h2(n) + h3(n)` always be admissible? Explain.
5. In your own words, explain how combining heuristics can improve search performance.


#### 1. Write the formula for combining them into a single heuristic that remains admissible.

To combine the three admissible heuristics `h1`, `h2`, and `h3` into a single heuristic that remains admissible, you can use the following formula:

```python
h_combined(n) = max(h1(n), h2(n), h3(n))
```

This formula takes the maximum value among the three heuristics for any given node `n`, ensuring thatthe combined heuristic remains admissible, as it will never overestimate the true cost to reach the goal.

#### 2. Why is this combination admissible?

The combination of admissible heuristics is admissible because the maximum of admissible heuristics is also admissible. This is because the maximum heuristic will never overestimate the true cost to reach the goal, as each individual heuristic is designed to be optimistic and not overestimate the cost. 

Therefore, using the maximum of `h1`, `h2`, and `h3` ensures that the combined heuristic remains admissible, as it will not overestimate the cost to reach the goal state.

#### 3. Why is taking the maximum usually better than choosing only one fixed heuristic?

Taking the maximum of multiple admissible heuristics is usually better than choosing only one fixed heuristic because it allows you to leverage the strengths of each heuristic. Each heuristic may provide different estimates for the cost to reach the goal, and by taking the maximum, you can ensure that you are using the most informative heuristic at each step of the search. 

This can lead to a more efficient search process, as it can help to prune more of the search space and guide the search towards the goal more effectively. In contrast, relying on a single fixed heuristic may result in suboptimal performance if that heuristic is not well-suited for certain parts of the search space. 

By combining multiple heuristics, you can create a more robust and effective search strategy that can adapt to different scenarios and provide better guidance towards the goal.

#### 4. Would taking the sum `h1(n) + h2(n) + h3(n)` always be admissible? Explain.

`No`, taking the sum `h1(n) + h2(n) + h3(n)` would not always be admissible. An admissible heuristic is one that never overestimates the true cost to reach the goal from node `n`. 

If each of the heuristics `h1`, `h2`, and `h3` is admissible, it means that each of them individually does not overestimate the cost. However, when you sum them up, the combined heuristic could potentially overestimate the true cost if the individual heuristics are not perfectly correlated or if they are not designed to complement each other. 

Therefore, while each heuristic may be admissible on its own, their sum may not necessarily be admissible.

#### 5. In your own words, explain how combining heuristics can improve search performance.

Combining heuristics can improve search performance by leveraging the strengths of each individual heuristic. When multiple admissible heuristics are available, taking the maximum of them ensures that we use the most informed estimate at each step of the search. This can lead to a more efficient search process, as it helps to prune more of the search space and guide the search towards the goal more effectively. 

Additionally, combining heuristics can provide a more robust and adaptive approach to search, as it allows the algorithm to adjust its behavior based on the characteristics of different parts of the search space. Overall, using a combination of heuristics can lead to faster convergence to the optimal solution and reduce the computational resources required for the search.

## Part C - Programming a Heuristic Search Solution

### Exercise 10. Implement Greedy Best-First Search and A* Search

In this exercise, you will solve a route-finding problem using heuristic search.

#### Problem
Use the Romania map below as a graph search problem.

Your tasks:

1. Implement **Greedy Best-First Search**.
2. Implement **A* Search**.
3. Use the straight-line distance to Bucharest as the heuristic.
4. Find a path from **Arad** to **Bucharest**.
5. Report:
   - the solution path,
   - the total path cost,
   - the number of expanded nodes.
6. Compare the outputs of the two algorithms.

#### Requirements
- Use a priority queue.
- Avoid infinite loops.
- Keep your code readable and modular.
- Write a short conclusion after running both algorithms.


In [1]:
from heapq import heappush, heappop

# Romania map: weighted graph
graph = {
    'Arad': {'Zerind': 75, 'Sibiu': 140, 'Timisoara': 118},
    'Zerind': {'Arad': 75, 'Oradea': 71},
    'Oradea': {'Zerind': 71, 'Sibiu': 151},
    'Sibiu': {'Arad': 140, 'Oradea': 151, 'Fagaras': 99, 'Rimnicu Vilcea': 80},
    'Timisoara': {'Arad': 118, 'Lugoj': 111},
    'Lugoj': {'Timisoara': 111, 'Mehadia': 70},
    'Mehadia': {'Lugoj': 70, 'Drobeta': 75},
    'Drobeta': {'Mehadia': 75, 'Craiova': 120},
    'Craiova': {'Drobeta': 120, 'Rimnicu Vilcea': 146, 'Pitesti': 138},
    'Rimnicu Vilcea': {'Sibiu': 80, 'Craiova': 146, 'Pitesti': 97},
    'Fagaras': {'Sibiu': 99, 'Bucharest': 211},
    'Pitesti': {'Rimnicu Vilcea': 97, 'Craiova': 138, 'Bucharest': 101},
    'Bucharest': {'Fagaras': 211, 'Pitesti': 101, 'Giurgiu': 90, 'Urziceni': 85},
    'Giurgiu': {'Bucharest': 90},
    'Urziceni': {'Bucharest': 85, 'Hirsova': 98, 'Vaslui': 142},
    'Hirsova': {'Urziceni': 98, 'Eforie': 86},
    'Eforie': {'Hirsova': 86},
    'Vaslui': {'Urziceni': 142, 'Iasi': 92},
    'Iasi': {'Vaslui': 92, 'Neamt': 87},
    'Neamt': {'Iasi': 87}
}

# Straight-line distance to Bucharest
heuristic = {
    'Arad': 366,
    'Bucharest': 0,
    'Craiova': 160,
    'Drobeta': 242,
    'Eforie': 161,
    'Fagaras': 176,
    'Giurgiu': 77,
    'Hirsova': 151,
    'Iasi': 226,
    'Lugoj': 244,
    'Mehadia': 241,
    'Neamt': 234,
    'Oradea': 380,
    'Pitesti': 100,
    'Rimnicu Vilcea': 193,
    'Sibiu': 253,
    'Timisoara': 329,
    'Urziceni': 80,
    'Vaslui': 199,
    'Zerind': 374
}

start = 'Arad'
goal = 'Bucharest'

#### Task 10.1 - Helper Functions

Write helper functions if needed, for example:
- a function to reconstruct the path,
- a function to compute the total path cost,
- a function to print results clearly.

In [ ]:
#Helper function
# to reconstruct the path from start to goal
def reconstruct_path(came_from: dict[str, str], 
                    current: str) -> list[str]:
    total_path = [current]
    while current in came_from:
        current = came_from[current]
        total_path.append(current)
    return total_path[::-1]  # reverse the path

# to compute the total cost of a path
def compute_cost(path: list[str]
, graph: dict[str, dict[str, int]]) -> int:
    return sum(graph[path[i]][path[i + 1]] 
        for i in range(len(path) - 1))

#to print results clearly
def print_report(name: str, path: list[str]
, cost: int, expanded: int) -> None:
    print(f"\n{'═'*50}")
    print(f"  {name}")
    print(f"{'═'*50}")
    print(f"  Path     : {' → '.join(path)}")
    print(f"  Cost     : {cost} km")
    print(f"  Expanded : {expanded} nodes")

#### Task 10.2 - Greedy Best-First Search

Implement Greedy Best-First Search using:

`f(n) = h(n)`

Your function should return at least:
- path
- total cost
- expanded nodes count

In [ ]:
#Greedy Best-First Search implementation
#using f(n) = h(n)
def greedy_best_first_search(graph, heuristic, start, goal):
    open_set = []
    heappush(open_set, (heuristic[start], start))
    came_from = {}
    expanded_nodes = set()

    while open_set:
        _, current = heappop(open_set)

        if current == goal:
            path = reconstruct_path(came_from, current)
            cost = compute_cost(path, graph)
            print_report("Greedy Best-First Search"
                , path, cost, len(expanded_nodes))
            return path

        expanded_nodes.add(current)

        for neighbor in graph[current]:
            if neighbor not in expanded_nodes:
                heappush(open_set
                    , (heuristic[neighbor], neighbor))
                if neighbor not in came_from:
                    came_from[neighbor] = current

    print_report("Greedy Best-First Search",
         [], 0, len(expanded_nodes))
    return None

#### Task 10.3 - A* Search

Implement A* Search using:

`f(n) = g(n) + h(n)`

Your function should return at least:
- path
- total cost
- expanded nodes count

In [ ]:
#implement A* Search
#using f(n) = g(n) + h(n)
def a_star_search(graph, heuristic, start, goal):
    open_set = []
    heappush(open_set, (heuristic[start], start))
    came_from = {}
    g_score = {node: float('inf') for node in graph}
    g_score[start] = 0
    expanded_nodes = set()

    while open_set:
        _, current = heappop(open_set)

        if current == goal:
            path = reconstruct_path(came_from, current)
            cost = compute_cost(path, graph)
            print_report("A* Search",
                 path, cost, len(expanded_nodes))
            return path

        expanded_nodes.add(current)

        for neighbor, cost in graph[current].items():
            tentative_g_score = g_score[current] + cost
            if tentative_g_score < g_score[neighbor]:
                came_from[neighbor] = current
                g_score[neighbor] = tentative_g_score
                f_score = tentative_g_score + heuristic[neighbor]
                heappush(open_set, (f_score, neighbor))

    print_report("A* Search", [], 0, len(expanded_nodes))
    return None

#### Task 10.4 - Run Both Algorithms

Run Greedy Best-First Search and A* Search from Arad to Bucharest.

Display the results in a readable format.

In [ ]:
greedy_best_first_search(graph, heuristic, start, goal)


══════════════════════════════════════════════════
  Greedy Best-First Search
══════════════════════════════════════════════════
  Path     : Arad → Sibiu → Fagaras → Bucharest
  Cost     : 450 km
  Expanded : 3 nodes


['Arad', 'Sibiu', 'Fagaras', 'Bucharest']

In [8]:
a_star_search(graph, heuristic, start, goal)


══════════════════════════════════════════════════
  A* Search
══════════════════════════════════════════════════
  Path     : Arad → Sibiu → Rimnicu Vilcea → Pitesti → Bucharest
  Cost     : 418 km
  Expanded : 5 nodes


['Arad', 'Sibiu', 'Rimnicu Vilcea', 'Pitesti', 'Bucharest']

#### Task 10.5 - Comparison and Reflection

Write a short comparison (5-8 sentences) answering:

1. Which algorithm found the better path?
2. Which algorithm seems more goal-directed?
3. Which algorithm is guaranteed to be optimal here, and why?
4. What role does the heuristic play in both algorithms?
5. What would happen if the heuristic overestimated too much?


##### 1. Which algorithm found the better path?

The A* Search found the better path with a cost of 418 km, while the Greedy Best-First Search found a path with a cost of 450 km. 

This is because A* considers both the cost to reach the current node and the estimated cost to reach the goal, while Greedy Best-First Search only considers the heuristic estimate to the goal.

##### 2. Which algorithm seems more goal-directed?

The A* Search seems more goal-directed because it takes into account the actual cost to reach each node and the estimated cost to reach the goal, making it more likely to find a path that leads directly toward the goal.

##### 3. Which algorithm is guaranteed to be optimal here, and why?

The A* Search is guaranteed to be optimal here because it uses an admissible heuristic, meaning it never overestimates the cost to reach the goal. This ensures that the first time it reaches the goal, it has found the optimal path.

##### 4. What role does the heuristic play in both algorithms?

In both algorithms, the heuristic guides the search process by providing an estimate of the cost from any given node to the goal. In A*, this estimate is used in conjunction with the actual cost to prioritize which nodes to explore next.

##### 5. What would happen if the heuristic overestimated too much?

If the heuristic overestimated too much, it could lead A* to explore suboptimal paths, potentially missing the true shortest path. In extreme cases, if the heuristic consistently overestimates by a large margin, A* might not find an optimal solution at all.


## Bonus Challenge (Optional)

Modify your A* implementation to support **Weighted A***:

`f(n) = g(n) + W * h(n)`

Try at least two values such as `W = 1.5` and `W = 2.0`.

Then discuss:
- how the number of expanded nodes changes,
- how the solution quality changes,
- whether the result is still guaranteed to be optimal.


In [ ]:
# Modify A* implementation to support Weighted A* Search
#using f(n) = g(n) + w * h(n)

def weighted_a_star_search(graph, heuristic, start, goal, weight=1.5):
    open_set = []
    heappush(open_set, (heuristic[start] * weight, start))
    came_from = {}
    g_score = {node: float('inf') for node in graph}
    g_score[start] = 0
    expanded_nodes = set()

    while open_set:
        _, current = heappop(open_set)

        if current == goal:
            path = reconstruct_path(came_from, current)
            cost = compute_cost(path, graph)
            print_report(f"Weighted A* Search (W={weight})",
                 path, cost, len(expanded_nodes))
            return path

        expanded_nodes.add(current)

        for neighbor, cost in graph[current].items():
            tentative_g_score = g_score[current] + cost
            if tentative_g_score < g_score[neighbor]:
                came_from[neighbor] = current
                g_score[neighbor] = tentative_g_score
                f_score = tentative_g_score + weight * heuristic[neighbor]
                heappush(open_set, (f_score, neighbor))

    print_report(f"Weighted A* Search (W={weight})",
         [], 0, len(expanded_nodes))
    return None


In [14]:
# W = 1.5
weighted_a_star_search(graph, heuristic, start, goal, weight=1.5)


══════════════════════════════════════════════════
  Weighted A* Search (W=1.5)
══════════════════════════════════════════════════
  Path     : Arad → Sibiu → Fagaras → Bucharest
  Cost     : 450 km
  Expanded : 3 nodes


['Arad', 'Sibiu', 'Fagaras', 'Bucharest']

In [15]:
# W = 2.0
weighted_a_star_search(graph, heuristic, start, goal, weight=2.0)


══════════════════════════════════════════════════
  Weighted A* Search (W=2.0)
══════════════════════════════════════════════════
  Path     : Arad → Sibiu → Fagaras → Bucharest
  Cost     : 450 km
  Expanded : 3 nodes


['Arad', 'Sibiu', 'Fagaras', 'Bucharest']

##### How the number of expanded nodes changes,
    
The number of expanded nodes generally decreases as the weight increases, because the algorithm becomes more greedy and relies more on the heuristic, which may lead to faster but less informed decisions.


##### How the solution quality changes,

As the weight increases, the solution quality may degrade, meaning that the path found may be longer than the optimal path. This is because the algorithm prioritizes nodes with lower heuristic values more heavily, which can lead to suboptimal paths.

##### Whether the result is still guaranteed to be optimal.

The result is no longer guaranteed to be optimal when the weight is greater than 1. This is because the algorithm may overlook better paths in favor of those that appear more promising based on the heuristic, leading to suboptimal solutions.